kotlin-codegen/compiler/src/main/java/com/squareup/moshi/Property.kt

```diff
@@ -0,0 +1,146 @@
+  }
+
+  fun allocateNames(nameAllocator: NameAllocator) {
+    localName = nameAllocator.newName(name)
+    localIsPresentName = nameAllocator.newName("${name}Set")
+    delegateName = nameAllocator.get(delegateKey())
+  }
+
+  /** Returns a key that matches keys of properties that can share an adapter. */
+  fun delegateKey() = unaliasedName to jsonQualifiers
+
+  /** Returns an adapter to use when encoding and decoding this property. */
+  fun generateDelegateProperty(enclosing: Adapter): PropertySpec {
+    val adapterTypeName = ParameterizedTypeName.get(
+        JsonAdapter::class.asTypeName(), unaliasedName)
+    val qualifiers = jsonQualifiers.toList()
+    val standardArgs = arrayOf(enclosing.moshiParam,
+        if (unaliasedName is ClassName && qualifiers.isEmpty()) {
+          ""
+        } else {
+          CodeBlock.of("<%T>", unaliasedName)
+        },
+        unaliasedName.makeType(
+            enclosing.elements, enclosing.typesParam, enclosing.genericTypeNames ?: emptyList()))
+    val standardArgsSize = standardArgs.size + 1
+    val (initializerString, args) = when {
+      qualifiers.isEmpty() -> "" to emptyArray()
+      qualifiers.size == 1 -> {
+        ", %${standardArgsSize}T::class.java" to arrayOf(
+            qualifiers.first().annotationType.asTypeName())
+      }
+      else -> {
+        val initString = qualifiers
+            .mapIndexed { index, _ ->
+              val annoClassIndex = standardArgsSize + index
+              return@mapIndexed "%${annoClassIndex}T::class.java"
+            }
+            .joinToString()
+        val initArgs = qualifiers
+            .map { it.annotationType.asTypeName() }
+            .toTypedArray()
+        ", $initString" to initArgs
+      }
+    }
+    val finalArgs = arrayOf(*standardArgs, *args)
+
+    return PropertySpec.builder(delegateName, adapterTypeName,
+        KModifier.PRIVATE)
+        .initializer("%1N.adapter%2L(%3L$initializerString)${if (nullable) ".nullSafe()" else ""}",
+            *finalArgs)
+        .build()
+  }
+
+  fun generateLocalProperty(): PropertySpec {
+    return PropertySpec.builder(localName, typeName.asNullable())
+        .mutable(true)
+        .initializer("null")
+        .build()
+  }
+
+  fun generateLocalIsPresentProperty(): PropertySpec {
+    return PropertySpec.builder(localIsPresentName, BOOLEAN)
+        .mutable(true)
+        .initializer("false")
+        .build()
+  }
+}
+
+/**
+ * Returns a suggested variable name derived from a list of type names.
+ */
+private fun List<TypeName>.toVariableNames(): String {
+  return joinToString("_") { it.toVariableName() }
+}
+
+/**
+ * Returns a suggested variable name derived from a type name.
+ */
+private fun TypeName.toVariableName(): String {
+  return when (this) {
+    is ClassName -> simpleName().decapitalize()
+    is ParameterizedTypeName -> {
+      rawType.simpleName().decapitalize() + if (typeArguments.isEmpty()) "" else "__" + typeArguments.toVariableNames()
```

Comment:
This underscore thing is super weird. The whole point of name allocator is to avoid tricks like this, no? Can fix in follow up.

Reply:
Will punt to follow-up. This PR doesn’t change behavior.